packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -145,42 +152,67 @@ const CustomElementHook: Hooks<VCustomElement> = {
             sel,
             owner: { renderer },
         } = vnode;
-        const elm = renderer.createElement(sel);
-
-        linkNodeToShadow(elm, vnode);
-        createViewModelHook(elm, vnode);
+        const CEBridge = registerTagName(sel, renderer);
+        (CEBridge as any).upgrade = (elm: HTMLElement) => {
+            createViewModelHook(elm, vnode);
+        };
+        const elm = new CEBridge();
         vnode.elm = elm;
+        linkNodeToShadow(elm, vnode);
 
-        allocateChildrenHook(vnode);
+        const vm = getAssociatedVMIfPresent(elm);
+        if (vm) {
+            allocateChildrenHook(vnode, vm);
+        }
         createCustomElmHook(vnode);
     },
     update: (oldVnode, vnode) => {
         updateCustomElmHook(oldVnode, vnode);
-        // in fallback mode, the allocation will always set children to
-        // empty and delegate the real allocation to the slot elements
-        allocateChildrenHook(vnode);
+        const vm = getAssociatedVMIfPresent(vnode.elm);
+        if (vm) {
+            // in fallback mode, the allocation will always set children to
+            // empty and delegate the real allocation to the slot elements
+            allocateChildrenHook(vnode, vm);
+        }
         // in fallback mode, the children will be always empty, so, nothing
         // will happen, but in native, it does allocate the light dom
         updateChildrenHook(oldVnode, vnode);
-        // this will update the shadowRoot
-        rerenderCustomElmHook(vnode);
+        if (vm) {
+            if (process.env.NODE_ENV !== 'production') {
+                assert.isTrue(
+                    isArray(vnode.children),
+                    `Invalid vnode for a custom element, it must have children defined.`
+                );
+            }
+            // this will update the shadowRoot
+            rerenderVM(vm);
```

Comment:
I don't know about this one... still trying to reasoning why we need re-render the vm after the update... this is tricky.

Reply:
we are good here after chatting offline.